File: leetcode3/남효정/3803. Count Residue Prefixes.py
class Solution:
    def residuePrefixes(self, s: str) -> int:
        ans = 0
        check = set()

        for i, char in enumerate(s):
            check.add(char)

            # 종류가 3가지 이상이 되면 더 계산할 필요 없음
            if len(check) >= 3:
                break
            
            # 문제에서 원하는 조건 만족할 때마다 정답 개수 += 1
            if len(check) == (i + 1) % 3:
                ans += 1
        return ans

File: leetcode3/염혜정/3803. Count Residue Prefixes.java
// 문자열의 0번째부터 set에 있는지 확인
// 없으면 넣어주고 distinct_cnt + 1
// O(n)

class Solution {
    public int residuePrefixes(String s) {
        Set<Character> set = new HashSet<>();
        int distinct_cnt = 0;
        int result = 0;
        for (int i = 0; i<s.length(); i++) {
            char c = s.charAt(i);
            if (!set.contains(c)) { // 중복이 아니라면
                distinct_cnt++;
                set.add(c);
            }
            int modulo = (i + 1) % 3;
            if (modulo == distinct_cnt) result++;
        }
        return result;
    }
}
